File: Python/Trabalho_POO_parte_2/Veiculo.py
# classe
from Roda import *

# biblioteca
from abc import ABC, abstractclassmethod, abstractmethod
#"ABC" significa "Abstract Base Class"
# classe abstrata
class Veiculo(ABC):
    # construtor abstrato
    @abstractclassmethod
    def __init__(self, ID: int, rodas: int):
        self.__id: int = ID
        self.__distancia_percorrida: int = 0
        self.__quantidade_rodas: int = rodas
        self.__rodas: list = [Roda() for _ in range(self.__quantidade_rodas)]

    # metodos get
    def get_id(self) -> int:
        return self.__id

    def get_distancia_percorrida(self) -> int:
        return self.__distancia_percorrida

    def get_quantidade_rodas(self) -> int:
        return self.__quantidade_rodas

    def get_rodas(self) -> list:
        return self.__rodas

    # metodos set
    def set_distancia_percorrida(self, new: int) -> int:
        self.__distancia_percorrida = new

    # funções abstratas
    @abstractmethod
    def mover(self) -> None:
        pass

    @abstractmethod
    def desenhar(self) -> None:
        pass

    # funções herdadas
    def verificar_pneu(self) -> bool:
        result = True
        for roda in self.__rodas:
            if not roda.get_calibragem():
                result = False
        return result

    def calibrar_pneu(self, id_pneu: int) -> None:
        self.__rodas[id_pneu].set_calibragem(True)
        print(f"Pneu: {id_pneu} -> calibrado")

    def esvaziar_pneu(self, id_pneu: int) -> None:
        self.__rodas[id_pneu].set_calibragem(False)
        print(f"Pneu: {id_pneu} -> esvaziado")

    # to_string = __str__
    def __str__(self) -> str:
        dados = []
        dados.append(f"\nID: {self.__id}")
        dados.append(f"Distancia percorrida: {self.__distancia_percorrida}")
        dados.append(f"Quantidade de rodas: {self.__quantidade_rodas}")
        for roda in self.__rodas:
            dados.append(str(roda))
        return "\n".join(dados)

File: C++/Trabalho_POO_parte_2/Veiculo.cpp
#include "Veiculo.h"


Veiculo::Veiculo(int ID, int q_rodas) : id(ID), quantidade_rodas(q_rodas), distancia_percorrida(0)
{
    for (int i = 0; i < quantidade_rodas; i++)
        rodas.push_back(Roda());
}

int Veiculo::get_id()
{
    return id;
}

int Veiculo::get_distancia_percorrida()
{
    return distancia_percorrida;
}

int Veiculo::get_quantidade_rodas()
{
    return quantidade_rodas;
}

vector<Roda> Veiculo::get_rodas()
{
    return rodas;
}

void Veiculo::set_distancia_percorrida(int new_distancia)
{
    distancia_percorrida = new_distancia;
}

bool Veiculo::verificar_pneu()
{
    for (Roda roda : rodas)
        if (!roda.get_calibragem())
            return false;
    return true;
}

void Veiculo::calibrar_pneu(int id_pneu)
{
    rodas[id_pneu].set_calibragem(true);
    std::cout << "Pneu: " << id_pneu << " -> calibrado" << endl;
}

void Veiculo::esvaziar_pneu(int id_pneu)
{
    rodas[id_pneu].set_calibragem(false);
    std::cout << "Pneu: " << id_pneu << " -> esvaziado" << endl;
}

string Veiculo::to_string()
{
    ostringstream oss;
    oss << "\nID: " << id << endl;
    oss << "Distancia percorrida: " << distancia_percorrida << endl;
    oss << "Quantidade de rodas: " << quantidade_rodas << endl;
    for (Roda roda : rodas)
        oss << roda.to_string();
    return oss.str();
}
